chess-AI/model_evaluation.py

```diff
@@ -0,0 +1,185 @@
+"""Helper file for evaluating quality of our learned models.
+
+These tests help us decide which models to keep during learning.
+"""
+import chess
+import torch
+import random
+import itertools
+import numpy as np
+
+from mcts import Mcts
+from ai_io import init_params
+from nn_layout import PlayNetwork
+from state_representation import get_cnn_input
+from output_representation import policy_converter
+from util import create_stockfish_wrapper
+
+
+class Knightr0Player:
+    '''"AI" class that is a wrapper around our custom modification of AlphaZero.
+    '''
+    def __init__(self, path_to_model=None):
+        if not(path_to_model == None):
+            self.model = torch.load(path_to_model)
+        else:
+            self.model = PlayNetwork(testing=True).to('cpu')
+        # TODO: Update exploration rate.
+        self.mcts = Mcts(exploration=0.01, device=torch.device('cuda:0' if torch.cuda.is_available() else 'cpu'))
+
+    def select_move(self, board):
+        '''Sets stockfish state from provided fen and returns best move.
+        '''
+        return Mcts.get_best_move(board.fen(), self.model)
+
+    def __str__(self):
+        return "Knightr0Player"
+
+
+class EasyStockfishPlayer:
+    """"AI" class that is a simple wrapper around the Stockfish engine.
+    """
+    def __init__(self, elo_rating=800):
+        self.stockfish = create_stockfish_wrapper()
+        self.stockfish.set_elo_rating(elo_rating)
+        self.stockfish.depth = 5
+
+    def select_move(self, board):
+        """Sets stockfish state from provided fen and returns best move.
+        """
+        self.stockfish.set_fen_position(board.fen())
+        return chess.Move.from_uci(self.stockfish.get_best_move())
+
+    def __str__(self):
+        return "EasyStockfishPlayer"
+
+
+class HardStockfishPlayer:
+    """"AI" class that is a simple wrapper around the Stockfish engine.
+    """
+    
+    def __init__(self, elo_rating=1800):
+        self.stockfish = create_stockfish_wrapper()
+        self.stockfish.set_elo_rating(elo_rating)
+        self.stockfish.depth = 9
+        self.elo = elo_rating
+
+    def select_move(self, board):
+        """Sets stockfish state from provided fen and returns best move.
+        """
+        self.stockfish.set_fen_position(board.fen())
+        return chess.Move.from_uci(self.stockfish.get_best_move())
+
+    def __str__(self):
+        return "HardStockfishPlayer"
+
+
+class RandomPlayer:
+    """"AI" class that is a simple wrapper around the Stockfish engine, generating random moves.
+    """
+    def select_move(self, board):
+        """Sets stockfish state from provided fen and returns random move.
+        """
+        # if type(move) is not str:
+        #     move = move['Move']
+        return random.choice(list(board.legal_moves))
+
+    def __str__(self):
+        return "RandomStockfishPlayer"
+        
+
+def evaluate_two_players(player1, player2, num_games=20):
+    white = False
```

Comment:
If white is just what player starts then shouldn't this just be random? `white = randint(0,1)` that way each game it is  just a random chance of either our AI starting or the enemy.

Reply:
I know I said I wouldn't be involved, but I'll chime in here to say I think there should be an equal number of games (can be a parameter that you guys figured out) of black and white games for each opponent. e.g. 5 games as white, 5 games as black vs. each opponent. That way the evaluation is less random since the way we compute the evaluation is the same. What do you guys think?

Reply:
That would work also. I think we need to track wins for white vs black. rather than just `wins=[0,0,0]` it should probably be `wins_white=[0,0,0]` and `wins_black=[0,0,0]`